st2common/tests/unit/test_config_loader.py

```diff
@@ -328,6 +332,199 @@ def test_get_config_dynamic_config_item_render_fails_user_friendly_exception_is_
+        Config.add_or_update(config_db)
+
+        expected_msg = ('Failed to render dynamic configuration value for key '
+                        '"level0_key" with value "{{ this is some invalid Jinja }}"'
+                        ' for pack ".*?" config: <class \'jinja2.exceptions.TemplateSyntaxError\'>'
+                        ' expected token \'end of print statement\', got \'Jinja\'')
+        self.assertRaisesRegexp(RuntimeError, expected_msg, loader.get_config)
+        config_db.delete()
+
+    def test_get_config_dynamic_config_item(self):
+        pack_name = 'dummy_pack_schema_with_nested_object_6'
+        loader = ContentPackConfigLoader(pack_name=pack_name)
+
+        ####################
+        # value in top level item
+        KeyValuePair.add_or_update(KeyValuePairDB(name='k1', value='v1'))
+        values = {
+            'level0_key': '{{st2kv.system.k1}}'
+        }
+        config_db = ConfigDB(pack=pack_name, values=values)
+        config_db = Config.add_or_update(config_db)
+
+        config_rendered = loader.get_config()
+
+        self.assertEquals(config_rendered, {'level0_key': 'v1'})
+
+        config_db.delete()
+
+    def test_get_config_dynamic_config_item_nested_dict(self):
+        pack_name = 'dummy_pack_schema_with_nested_object_7'
+        loader = ContentPackConfigLoader(pack_name=pack_name)
+
+        KeyValuePair.add_or_update(KeyValuePairDB(name='k0', value='v0'))
+        KeyValuePair.add_or_update(KeyValuePairDB(name='k1', value='v1'))
+        KeyValuePair.add_or_update(KeyValuePairDB(name='k2', value='v2'))
+
+        ####################
+        # values nested dictionaries
+        values = {
+            'level0_key': '{{st2kv.system.k0}}',
+            'level0_object': {
+                'level1_key': '{{st2kv.system.k1}}',
+                'level1_object': {
+                    'level2_key': '{{st2kv.system.k2}}'
+                }
+            }
+        }
+        config_db = ConfigDB(pack=pack_name, values=values)
+        config_db = Config.add_or_update(config_db)
+
+        config_rendered = loader.get_config()
+
+        self.assertEquals(config_rendered,
+                          {
+                              'level0_key': 'v0',
+                              'level0_object': {
+                                  'level1_key': 'v1',
+                                  'level1_object': {
+                                      'level2_key': 'v2'
+                                  }
+                              }
+                          })
+
+        config_db.delete()
+
+    def test_get_config_dynamic_config_item_list(self):
+        pack_name = 'dummy_pack_schema_with_nested_object_7'
+        loader = ContentPackConfigLoader(pack_name=pack_name)
+
+        KeyValuePair.add_or_update(KeyValuePairDB(name='k0', value='v0'))
+        KeyValuePair.add_or_update(KeyValuePairDB(name='k1', value='v1'))
+
+        ####################
+        # values in list
+        values = {
```

Comment:
Ideally we would also have some tests which verify schema is correctly handled for arrays items (aka schema which defines items.type or similar).

Reply:
I agree. As a stop-gap i was trying to at least include some tests to cover various structural corner cases in the config.

Reply:
I will merge this, but if you get a chance it would be great if you can open a PR with tests for that scenario :)